apps/openmw/mwworld/inventorystore.cpp

```diff
@@ -363,161 +535,79 @@ void MWWorld::InventoryStore::autoEquip (const MWWorld::Ptr& actor)
-        {
-            std::pair<std::vector<int>, bool> itemsSlots =
-                weapon->getClass().getEquipmentSlots (*weapon);
-
-            if (!itemsSlots.first.empty())
-            {
-                if (!itemsSlots.second)
-                {
-                    if (weapon->getRefData().getCount() > 1)
-                    {
-                        unstack(*weapon, actor);
-                    }
-                }
-
-                int slot = itemsSlots.first.front();
-                slots_[slot] = weapon;
-            }
-
-            break;
-        }
-
-        weaponSkillVisited[maxWeaponSkill] = true;
-    }
-
-    bool changed = false;
-
-    for (std::size_t i=0; i<slots_.size(); ++i)
-    {
-        if (slots_[i] != mSlots[i])
-        {
-            changed = true;
-            break;
-        }
-    }
-    mUpdatesEnabled = true;
-
-    if (changed)
-    {
-        mSlots.swap (slots_);
-        fireEquipmentChangedEvent(actor);
-        updateMagicEffects(actor);
-        flagAsModified();
-    }
 }
 
-void MWWorld::InventoryStore::autoEquipShield(const MWWorld::Ptr& actor)
+void MWWorld::InventoryStore::autoEquipShield(const MWWorld::Ptr& actor, TSlots& slots_)
 {
-    bool updated = false;
-
-    mUpdatesEnabled = false;
     for (ContainerStoreIterator iter(begin(ContainerStore::Type_Armor)); iter != end(); ++iter)
     {
         if (iter->get<ESM::Armor>()->mBase->mData.mType != ESM::Armor::Shield)
             continue;
-
         if (iter->getClass().canBeEquipped(*iter, actor).first != 1)
             continue;
-
         if (iter->getClass().getItemHealth(*iter) <= 0)
             continue;
-
         std::pair<std::vector<int>, bool> shieldSlots =
             iter->getClass().getEquipmentSlots(*iter);
-
         if (shieldSlots.first.empty())
             continue;
-
         int slot = shieldSlots.first[0];
         const ContainerStoreIterator& shield = mSlots[slot];
-
         if (shield != end()
                 && shield.getType() == Type_Armor && shield->get<ESM::Armor>()->mBase->mData.mType == ESM::Armor::Shield)
         {
             if (shield->getClass().getItemHealth(*shield) >= iter->getClass().getItemHealth(*iter))
                 continue;
         }
+        slots_[slot] = iter;
+    }
+}
+
+void MWWorld::InventoryStore::autoEquip (const MWWorld::Ptr& actor)
+{
+    TSlots slots_;
+    initSlots (slots_);
 
-        equip(slot, iter, actor);
-        updated = true;
+    // Disable model update during auto-equip
+    mUpdatesEnabled = false;
+
+    bool weaponSheathing = Settings::Manager::getBool("weapon sheathing", "Game");
+    if (weaponSheathing)
+    {
+        // We should not auto-equip weapons and ammo if actor is a trader and sells them
+        int services = actor.getClass().getServices(actor);
+        bool sellsWeapon = services & (ESM::NPC::Weapon|ESM::NPC::MagicItems);
+        if (!sellsWeapon)
+            autoEquipWeapon(actor, slots_);
```

Comment:
Another regression? Why is this limited to weapon sheathing being enabled?

Reply:
IIRC, auto-equipping weapons is desired even if they are not shown visually. They still have impact on gameplay - in particular, you should not be able to pickpocket equipped weapons. I think there is a (resolved) bug report regarding that but I can't find it right now.

Reply:
Ok, I'll change it in this case.

Reply:
For now I use weapon and ammo autoequipping for all NPCs, as in Morrowind, also for creatures which not sell weapons. Is it OK?

Reply:
Sounds good to me.